Implement initial Sai2 support

Wunkolo opened the pull request:
Implements initial support for `.sai2` file format.

Addresses https://github.com/Wunkolo/libsai/issues/14

Wunkolo committed:
Add Sai2 thumbnail extraction sample

Doesn't work just yet at the moment. Intended as a test-bed of sorts for now to extract singular tiles

Wunkolo committed:
Init Sai2 header

Migrate some foundational types and enums to the new sai2 header.

Wunkolo committed:
Fix Thumbnail-Sai2 translation unit

Wunkolo committed:
Refactor Sai2 Canvas file types

Wunkolo committed:
Add `Tag{LE,BE}` specifiers

Wunkolo committed:
Update clang format

Primarily changes the column width to 80

Wunkolo committed:
Refactor RLE decompression iteration code

Passes more files _except_ for the vertical-stripes sample file.

Wunkolo committed:
Fix tile iteration

Using std::uint8_t felt right, but it wraps around at the 256th iteration causing iteration errors. Now all sample files are being read correctly

Wunkolo committed:
Fix signed/unsigned comparison

This is a build-blocking error on windows

Wunkolo committed:
Draft `DeltaUnpack16`

Converting the sse instructions here into intrinsics to simulate the original implementation.
Seems to work well enough, though when rows are the same(such as when the tile is filled completely with the color red or with a vertical strike pattern) the subsequent rows seem to just be zeroes.
There is some optimization happening somewhere that allows duplicate rows of tile data to not have to be stored again.

Wunkolo committed:
Implement delta-row decompression

This now extracts tiles properly!

Wunkolo committed:
Fix tile dump file path type

Breaks on windows due to possible UTF encoding.

Wunkolo committed:
Add generic implementation of `UnpackDeltaRLE16`

Implements a generic implementation of this function based on the original SSE intrinsics.
This can be conditionally re-SIMDed later.
Enables MacOS/Windows Arm64 builds again and x86 builds.

Wunkolo committed:
Fix big-endian tag values, add init "old" thumbnail format

Wunkolo committed:
Init Fssj data iteration

Seems to be iterated in blocks of 4096, with cases in which the data _may_ be smaller. Still not sure what the data is intended to represent exactly.

Wunkolo committed:
Add `up` PNG Filter finding

This delta-previous-row function looks like an implementation of the PNG specification's "up" filter mode, where the values are just the compressed differences from the previous row.

Wunkolo committed:
Remove unused BlockBuffer

Wunkolo commented:
Scanlines of a `dpcm` tile seem to implement something similar to PNG's `Up` filter seen here:
http://www.libpng.org/pub/png/spec/1.2/PNG-Filters.html

The first scanline of a tile is always stored literally, and rows after will store deltas pixel-values of the previous row. This allows tiles to compress more optimally if the pixel-contents are flag regions of the same color(a similar benefit of PNG's compression).

Wunkolo committed:
Add sample `fssj` block observations

photopea commented:
Hey, I have written this "partial" SAI2 format specification, i.e. "what I know so far" - https://github.com/photopea/SAI2-specification/ , but maybe you know all this.

The number of tiles of a layer (C) is stored in the "layr" chunk.

Wunkolo committed:
Update sai2 BlobData endian

`dpcm` makes the most sense as these are "delta compressed"

Wunkolo committed:
Refactor `FssJ`->`Jssf`

Still not sure what this format is exactly. Jpeg/Jfif tiles?

Wunkolo committed:
Draft `DecompressFssjBlock`

Using the original assembly directly here

Wunkolo committed:
Refactor `ConvertJssfToRGB`

This is still seemingly not complete.
There are some intermediate operations passes done before/after this conversion process it seems.

Wunkolo committed:
Refactor `ConvertRGB8ToYUV16`

This is the weird color-space that Sai2 seems to use for JSSF blocks. Compression still needs to be figured out.

Wunkolo committed:
Add sai2 translation unit

Wunkolo committed:
Implement `IterateCanvasData`

Central utility function for iterating all canvas-file data

Wunkolo committed:
Refactor delta-compressed data into sai2 translation unit

Move more utility functions out of the thumbnail sample

photopea commented:
Wow, did you manage to understand the SAI2 pixel compression?

Could you maybe describe it in a natural language, so that I can write it into https://github.com/photopea/SAI2-specification/ ? Also, did you check my specification? Did you know all this?

Wunkolo commented:
There are multiple formats used for thumbnails: `jssf` and `dpcm`. `jssf` seems to be used in older versions of Sai2 in particular and seems to be very complex, and involves an intermediate color-space conversion. I am still figuring this out.

`dpcm`, used by newer versions of sai, is basically figured out but it's due for some more testing. The main function for that is here:
https://github.com/Wunkolo/libsai/pull/20/files#diff-e9cda01de5c5da7982099efa30ae90dda40a0921d1add9917f8aef2c89307a0dR242-R361
Each tile is compressed, and reach row within a tile is delta-compressed using the previous row. The delta-compressed data itself is also compressed using the a modified Run-Length-Encoding scheme seen here:
https://github.com/Wunkolo/libsai/pull/20/files#diff-02d0aedaa1fda248512c60cecfb1fdc85680b0762355bee688058622cd6748c2R164-R311

I am still working on this in my spare time. Trying to document it all in plain-English for someone else while I am still researching and experimenting would slow me down at the moment. When I do care to document it in plain-English, it would just be within this repo similar to [this one](https://github.com/Wunkolo/libsai/blob/main/README.md).

For now, your best bet would be to [check my notes in the actual code](https://github.com/Wunkolo/libsai/pull/20/files) to follow along if you wish until I start drafting up a specification.

Wunkolo committed:
FIx double->float truncation

Fixes windows builds

Mickommic commented:
Here a [viewer](http://mickommicg.temp.swtest.ru/MickView.zip) with support for sai and sai2 files. Bugs, lags and crashes are quite likely.

photopea commented:
@Wunkolo  I guess that @Mickommic made a Github account today to ask us to download a virus? :D

Mickommic commented:
Try it on a virtual machine or in a sandbox. :D

Mickommic commented:
Hello!
I want to thank Wunkolo for his description of sai and sai2. My respect to you.
Can you help me with this?

```
function Sai2ReadThumb(Dst: Pointer; Stream: TStream; Info: PiSai2Info; PixelStep, LineStep: Inteher): Boolean;
var
  i, j, w, h: Integer;
  c: Word;
begin
  Result:= False;

  if (Dst = nil) or (Stream = nil)
  or (Info = nil) or (Info^.ThumbOffset = 0) then Exit;

  Stream.Position:= Info^.ThumbOffset;
  Stream.Read(w, 4);//width
  Stream.Read(h, 4);//height
  Stream.Read(i, 4);

  if TiFourChar(i) <> 'jssf' then Exit;

  Stream.Read(i, 2);//width again
  Stream.Read(j, 2);//height again

  if (Word(i) <> Word(w)) or (Word(j) <> Word(h)) then Exit;

  Stream.Read(c, 2);//channels?
(*
then there are 128 bytes of data repeated in each file

02 02 02 02 02 02 03 03 03 03 04 04 04 04 04 04
04 04 04 04 04 06 06 06 06 06 06 06 06 06 06 06
06 06 06 06 08 08 08 08 08 08 08 09 09 09 09 09
09 0A 0A 0A 0A 0A 0C 0C 0C 0C 0E 0E 0E 10 10 10
06 06 06 06 06 06 08 08 08 08 09 09 09 09 09 0C
0C 0C 0C 0C 0C 10 10 10 10 10 10 10 13 13 13 13
13 13 13 13 17 17 17 17 17 17 17 1B 1B 1B 1B 1B
1B 20 20 20 20 20 25 25 25 25 25 25 25 25 25 25
*)


//I don't know how to interpret this



  Exit;


  Result:= True;
end;

```

Mickommic commented:
> > Can you help me with this?
> 
> [I am still deciphering `jssf` data in particular at the moment and you can see the state of it here](https://github.com/Wunkolo/libsai/pull/20/files#diff-e9cda01de5c5da7982099efa30ae90dda40a0921d1add9917f8aef2c89307a0dR189). This makes two people that want me to contribute to their own project's code-base... I'd rather people just follow the code I am writing so I am not bifurcated across several projects.

Okay, I got it. Here is the [source code](http://mickommicg.temp.swtest.ru/Mickom.sai2.zip) of my sai2 file decoder written according to your descriptions. Maybe it will be useful to you. And many thanks to you and everyone who helps you for your work.

Mickommic commented:
Hello!
I believe that jssf is a JPEG stream without any jpeg headers, except for two quantization tables.
One table for the luminance channel, 
```
02 02 02 02 02 02 03 03
03 03 04 04 04 04 04 04
04 04 04 04 04 06 06 06
06 06 06 06 06 06 06 06
06 06 06 06 08 08 08 08
08 08 08 09 09 09 09 09
09 0A 0A 0A 0A 0A 0C 0C
0C 0C 0E 0E 0E 10 10 10

```
the other for the chrominance channels.
```
06 06 06 06 06 06 08 08
08 08 09 09 09 09 09 0C
0C 0C 0C 0C 0C 10 10 10
10 10 10 10 13 13 13 13
13 13 13 13 17 17 17 17
17 17 17 1B 1B 1B 1B 1B
1B 20 20 20 20 20 25 25
25 25 25 25 25 25 25 25

```
I tried to decode the bitstream using the typical Huffman tables defined in the [JPEG specification](https://www.w3.org/Graphics/JPEG/itu-t81.pdf). But it didn't work. I believe that SYSTEMAX has predefined its own Huffman tables, which should be contained in the exe in the constants section.

Wunkolo commented:
This is what I am suspect of as well, even just based on the `jssf`-name alone and some of the subroutines around reading/writing. If the user's system has more than 1 core, then it seems to dispatch threads that work in 32x32-pixel tiles, indicating that these 4096-byte(or less) chunks of data are likely a specialized jfif stream like you describe. It's possible it's storing DCT-coefficients for 32x32 tiles, or a 4x4 tile of 8x8 tiles like the jpeg standard.

Wunkolo commented:
> These are 8x8 block encoding procedures with ForwardDCT, Quantization and Huffman coding
Thanks for that, I'm finding more implementation-details now and am finding that it _is_ doing a similar jpeg-like implementation with the nuance that almost all arithmetic is sustained as 16-bit arithmetic, including all the look-up tables and constants and such which diverges from the usual JPEG implementations you would find online. The implementation writes `SYSTEMAX JFIF Encoder Ver.2.0` when writing to a regular 8-bit `jpeg` file and shares a lot of common functions and tables before doing a final 8-bit conversion. The main jssf decoding thread dispatch seems to be at `0x140204570` for me.

Mickommic commented:
Hello!
Finally figured out JSSF. Before the bitstream of each MCU-row there is its length (2 bytes).
That is:
length, bitstream of the first MCU-row, length, bitstream of the second MCU-row and so on.
<details><summary>Sai2ReadThumb</summary>
<p>

```
function Sai2ReadThumb(Dst: Pointer; Stream: TStream; Info: PiSai2Info; PixelStep, LineStep: Inteher): Boolean;
label Quit;
const                        //id  hsf/vsf  qt  id  hsf/vsf  qt  id  hsf/vsf  qt
  FFC0: array[0..8] of Byte = (1,   $11,    0,  2,   $11,    1,  3,   $11,    1);

                            //id   ht   id   ht   id   ht
  FFDA: array[0..5] of Byte = (1,   0,   2,  $11,  3,  $11);

(* uncomment this if your jpeg decoder doesn't know about the existence of typical Huffman tables
                            //     DHT     length
  FFC4: array[0..419] of Byte = ($ff,$c4,  1,$a2,
//   id ||          16 code length counters               ||   codes, the number of which is the sum of the length counters
      0,{}0, 1, 5, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, {}0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11,
      1,{}0, 3, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, {}0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11,
    $10,{}0, 2, 1, 3, 3, 2, 4, 3, 5, 5, 4, 4, 0, 0, 1,125,{}$01,$02,$03,$00,$04,$11,$05,$12,$21,$31,$41,$06,$13,$51,$61,$07,$22,$71,$14,$32,$81,$91,$A1,$08,$23,$42,$B1,$C1,$15,$52,$D1,$F0,$24,$33,$62,$72,$82,$09,$0A,$16,$17,$18,$19,$1A,$25,$26,$27,$28,$29,$2A,$34,$35,$36,$37,$38,$39,$3A,$43,$44,$45,$46,$47,$48,$49,$4A,$53,$54,$55,$56,$57,$58,$59,$5A,$63,$64,$65,$66,$67,$68,$69,$6A,$73,$74,$75,$76,$77,$78,$79,$7A,$83,$84,$85,$86,$87,$88,$89,$8A,$92,$93,$94,$95,$96,$97,$98,$99,$9A,$A2,$A3,$A4,$A5,$A6,$A7,$A8,$A9,$AA,$B2,$B3,$B4,$B5,$B6,$B7,$B8,$B9,$BA,$C2,$C3,$C4,$C5,$C6,$C7,$C8,$C9,$CA,$D2,$D3,$D4,$D5,$D6,$D7,$D8,$D9,$DA,$E1,$E2,$E3,$E4,$E5,$E6,$E7,$E8,$E9,$EA,$F1,$F2,$F3,$F4,$F5,$F6,$F7,$F8,$F9,$FA,
    $11,{}0, 2, 1, 2, 4, 4, 3, 4, 7, 5, 4, 4, 0, 1, 2,119,{}$00,$01,$02,$03,$11,$04,$05,$21,$31,$06,$12,$41,$51,$07,$61,$71,$13,$22,$32,$81,$08,$14,$42,$91,$A1,$B1,$C1,$09,$23,$33,$52,$F0,$15,$62,$72,$D1,$0A,$16,$24,$34,$E1,$25,$F1,$17,$18,$19,$1A,$26,$27,$28,$29,$2A,$35,$36,$37,$38,$39,$3A,$43,$44,$45,$46,$47,$48,$49,$4A,$53,$54,$55,$56,$57,$58,$59,$5A,$63,$64,$65,$66,$67,$68,$69,$6A,$73,$74,$75,$76,$77,$78,$79,$7A,$82,$83,$84,$85,$86,$87,$88,$89,$8A,$92,$93,$94,$95,$96,$97,$98,$99,$9A,$A2,$A3,$A4,$A5,$A6,$A7,$A8,$A9,$AA,$B2,$B3,$B4,$B5,$B6,$B7,$B8,$B9,$BA,$C2,$C3,$C4,$C5,$C6,$C7,$C8,$C9,$CA,$D2,$D3,$D4,$D5,$D6,$D7,$D8,$D9,$DA,$E2,$E3,$E4,$E5,$E6,$E7,$E8,$E9,$EA,$F2,$F3,$F4,$F5,$F6,$F7,$F8,$F9,$FA
  );
*)
var
  i, j, k: Integer;
  c, h, w: SmallInt;
  buf: PiByteArray;
  pjd: PiJpegDec;
  tid: TiImageData;
begin
  Result:= False;

  if (Dst = nil) or (Stream = nil)
  or (Info = nil) or (Info^.ThumbOffset = 0) then Exit;

  Stream.Position:= Info^.ThumbOffset;
  Stream.Read(i, 4);//width
  Stream.Read(j, 4);//height
  Stream.Read(k, 4);

  if TiFourChar(k) <> 'jssf' then Exit;

  Stream.Read(w, 2);//width again
  Stream.Read(h, 2);//height again

  if (w <> i) or (h <> j) then Exit;

  Stream.Read(c, 2);//channels

  pjd:= nil;
  buf:= nil;
  k:= Info^.ThumbLength - (Ord(c > 1) * 64) - 82;
  i:= k + 180;
(* uncomment this if your jpeg decoder doesn't know about the existence of typical Huffman tables
  Inc(i, SizeOf(FFC4));
*)
  try
    GetMem(buf, i);
  except
    Exit;
  end;

  pjd:= JpegDecodeInit;

  if pjd = nil then goto Quit;

  PWord(buf)^:= $d8ff;//SOI - Start of image
  PWord(@buf^[2])^:= $dbff;//DQT - Define quantization table
  PWord(@buf^[4])^:= WordSwap(Ord(c > 1) * 65 + 67);
  buf^[6]:= 0;//id of first QT for luma

  if Stream.Read(buf^[7], 64) < 64 then goto Quit;//reading first QT

  i:= 71;

  if c > 1 then begin
    buf^[71]:= 1;//id of second QT for chroma

    if Stream.Read(buf^[72], 64) < 64 then goto Quit;//reading second QT

    Inc(i, 65);
  end;
(* uncomment this if your jpeg decoder doesn't know about the existence of typical Huffman tables
  MoveMem(@buf^[i], @FFC4, SizeOf(FFC4));//DHT - Define Huffman table
  Inc(i, SizeOf(FFC4));
*)
  PWord(@buf^[i])^:= $c0ff;//SOFn - Start of frame   (SOF0 - Baseline DCT)
  PWord(@buf^[i + 2])^:= WordSwap(c * 3 + 8);
  buf^[i + 4]:= 8;//bps
  PWord(@buf^[i + 5])^:= WordSwap(h);
  PWord(@buf^[i + 7])^:= WordSwap(w);
  buf^[i + 9]:= c;//channels

  MoveMem(@buf^[i + 10], @FFC0, c * 3);
  Inc(i, c * 3 + 10);

  PInteger(@buf^[i])^:= $400ddff;//DRI - Define restart interval
  PWord(@buf^[i + 4])^:= WordSwap((w + 7) shr 3);
  Inc(i, 6);

  PWord(@buf^[i])^:= $daff;//SOS - Start of Scan
  PWord(@buf^[i + 2])^:= WordSwap(c * 2 + 6);
  buf^[i + 4]:= c;//channels

  MoveMem(@buf^[i + 5], @FFDA, c * 2);
  Inc(i, c * 2 + 5);

  PInteger(@buf^[i])^:= $3f00;//Ss, Se, Ah/Al
  Inc(i, 3);
  w:= 0;

  if Stream.Read(c, 2) < 2 then goto Quit;//length of the bit stream of the first MCU-row

  repeat
    j:= c + 2;

    if Stream.Read(buf^[i], j) < j then Break;//reading jpeg bitstream (entropy)

    Inc(i, j);
    c:= PWord(@buf^[i - 2])^;//length of the bit stream of the next MCU-row
    PWord(@buf^[i - 2])^:= ((w and 7) shl 8) or $d0ff;//RSTm - Restart with modulo 8 count “m”
    Inc(w);
    Dec(h, 8);
  until h <= 0;

  PWord(@buf^[i - 2])^:= $d9ff;//EOI - End of image

//At this point feed buf to your preferred jpeg decoder

  k:= i;

  tid.Component[0]:= Dst;
  tid.Component[1]:= @PiByteArray(Dst)^[1];
  tid.Component[2]:= @PiByteArray(Dst)^[2];
  tid.SampleStep[0]:= PixelStep;
  tid.SampleStep[1]:= PixelStep;
  tid.SampleStep[2]:= PixelStep;
  tid.LineStep[0]:= LineStep;
  tid.LineStep[1]:= LineStep;
  tid.LineStep[2]:= LineStep;
  tid.BitDepth:= 8;

  repeat
    j:= JpegDecode(@buf^[k - i], i, pjd);

    if pjd^.Scan.DecodedLines.Count <> 0 then begin
      JpegDecodeComponents(tid, pjd);

      Inc(Inteher(tid.Component[0]), LineStep shl 3);
      Inc(Inteher(tid.Component[1]), LineStep shl 3);
      Inc(Inteher(tid.Component[2]), LineStep shl 3);
    end;
  until (j < 0) or (j > 4) or (i <= 0);

  Result:= True;
Quit:
  JpegDecodeFree(pjd);
  FreeMem(buf);
end;
``` 

</p>
</details>

Wunkolo committed:
Init `jssf` MCU decoding

Wunkolo committed:
Update project to C++23

Mostly for `std::byteswap` but this can be reduced again later with some local versions of this function.

Wunkolo committed:
Draft converting `jssf` into standard jpeg stream

Wunkolo committed:
Add `stb_image`

Used to convert the generated jpeg stream into a regular RGB8 image

Wunkolo committed:
Draft direct jpeg output

stb_image does not properly handle `RSTm` markers at the moment, but outputting a jpeg stream seems to work just fine!

Wunkolo commented:
Thanks for this insight, learned a lot about jpeg encoding from this.
I was able to extract `jssf` data directly into a jpeg stream by adding the appropriate markers and metadata around the entropy bytes. The next step would be to decode these thumbnails locally and get a regular `RGB8`-buffer so that it can be more generally useful. Currently, simpler jpeg decoders like [stb_image](https://github.com/nothings/stb/blob/master/stb_image.h) do not properly support the `RSTm` marker and won't decode it properly. But it might be viable to implement RSTm support and sustain the changes locally.

Here is an extracted thumbnail from a sai2 document.
![gal-thumbnail](https://github.com/user-attachments/assets/50281538-53d5-4ac8-9ac8-a1c282eb1fa2)
Because it uses `RSTm` markers, a lot of jpg-decoders are not able to view or decode this image properly(except maybe the compliant jpeg-decoder that your web browser uses if you can see the image). If you try to save this image, you'll find some image-viewers will not be able to open this image properly such as Windows Photos or MSPaint or XnView, but this is still a great starting point now that both of the primary thumbnail formats have been mapped out.

Wunkolo committed:
Remove `ConvertRGB8ToYUV16`

No need to manually handle jpeg decoding here anymore

Wunkolo committed:
Refactor for C++20 features

Remove C++23 requirement and allow C++20 features. Primarily removing the need for `std::byteswap` but just explicitly handling the element types. Should get MacOS builds working again too.

Wunkolo committed:
Fix Huffman table interpretation

Properly inject the DHT data so that `stb_image` can properly decode the intermediate jpeg

Wunkolo committed:
Add additional Huffman tables when thumbnail is color

Wunkolo committed:
Remove final `RSTm` marker

No need to restart the decoder when the image is about to end right after.

Wunkolo committed:
Refactor thumbnail extraction prototype

Making it similar to the original libsai interface where a weird tuple is being returned to remove the image-writing procedure out of the extraction code.

Delta-Compressed thumbnails still needs to be figured out for thumbnails larger than 1 tile at the moment though.

Wunkolo committed:
Fix table entry printf FourCC type

Wunkolo committed:
Refactor Jssf->Jpeg conversion to `ExtractJssfToJpeg`

This will soon be put into the Sai2 library namespace. trying to avoid LibSai having a hard dependency on a jpeg decoder, ideally.

Wunkolo committed:
Refactor Jssf to Jpeg conversion into Sai2 library

Wunkolo committed:
Quit at the first found supported thumbnail

At this point, the older thumbnail format is proving to be quite rare, while the jpeg-stream based format is so much more common in my small "survey". So I think it's worth doing a public release with just the jssf format for now and supporting the older png-filter based one later.

Wunkolo committed:
Refactor Jssf extraction routines into Sai2 library

Wunkolo committed:
Add missing `pragma once`

Wunkolo committed:
Add additional canvas data tags

Wunkolo committed:
Add Canvas Background flag data

This is also used to determine the number of channels for the newer thumbnail data

Wunkolo committed:
Add initial `layr` and `hist` structures

Wunkolo committed:
Fix `dpcm` canvas background flag compile error

Wunkolo committed:
Add `raw` data tag

Seems primarily unused, but worth having

Wunkolo committed:
Refactor Delta-Encoded tile iteration

Address into a global image, and read the tile-size array before iterating the tile-data-decompression routine itself. Also ensures the "Checksums" are valid while iterating.

Wunkolo committed:
Fix unaligned RLE bitstream support

Additional edge-case handling that Sai does not do when reading from an RLE bitstream that is not aligned to 4-byte elements.

Wunkolo committed:
Utilize tile-size table for iteration

Things _seem_ to iterate fine, save for the strange opcode issue in the RLE stream now.

Wunkolo committed:
Fix tile checksum iteration

These checksum values are for each column, and each row

Wunkolo committed:
Fix Delta-Compressed indexing

The image may not be a factor of the tile-size, so `Width` must be used here. And with that, images are successfully extracting!!!

Wunkolo committed:
Fix RLE bitstream iteration

Wunkolo committed:
Refactor `dpcm` extraction into sai2 library
